File: Languages/python/algorithms-dsa/leetcode/monotonic-stack/next_greater_element.py
"""
LeetCode #496 - Next Greater Element I
Topic: Monotonic Stack
Difficulty: Easy

Given two arrays nums1 (subset of nums2), for each element in nums1 find the
next greater element in nums2. Return -1 if none exists.

Approach: Use a decreasing monotonic stack to precompute next greater element
for every value in nums2, store in hash map. Then look up for nums1.

Time Complexity: O(n + m)
Space Complexity: O(n)
"""


def next_greater_element(nums1: list[int], nums2: list[int]) -> list[int]:
    # Build next-greater map from nums2
    next_greater: dict[int, int] = {}
    stack: list[int] = []  # decreasing stack of values

    for num in nums2:
        while stack and stack[-1] < num:
            next_greater[stack.pop()] = num
        stack.append(num)

    # Remaining elements have no next greater
    return [next_greater.get(num, -1) for num in nums1]


if __name__ == "__main__":
    assert next_greater_element([4, 1, 2], [1, 3, 4, 2]) == [-1, 3, -1]
    assert next_greater_element([2, 4], [1, 2, 3, 4]) == [3, -1]
    assert next_greater_element([1], [1]) == [-1]
    assert next_greater_element([1, 3, 5], [6, 5, 4, 3, 2, 1, 7]) == [7, 7, 7]
    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/leetcode/monotonic-stack/next_greater_element.cpp
/**
 * LeetCode 496: Next Greater Element I
 * Topic: Monotonic Stack
 * Difficulty: Easy
 *
 * Find the next greater element for each element in nums1 from nums2.
 * Use a decreasing monotonic stack to process nums2 right-to-left (or left-to-right with pop).
 * Time: O(n + m), Space: O(n)
 */
#include <iostream>
#include <vector>
#include <stack>
#include <unordered_map>
#include <cassert>
using namespace std;

vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
    unordered_map<int, int> nextGreater;
    stack<int> st; // decreasing stack

    for (int num : nums2) {
        while (!st.empty() && st.top() < num) {
            nextGreater[st.top()] = num;
            st.pop();
        }
        st.push(num);
    }

    vector<int> result;
    for (int num : nums1) {
        result.push_back(nextGreater.count(num) ? nextGreater[num] : -1);
    }
    return result;
}

int main() {
    vector<int> n1 = {4, 1, 2}, n2 = {1, 3, 4, 2};
    assert((nextGreaterElement(n1, n2) == vector<int>{-1, 3, -1}));

    vector<int> n3 = {2, 4}, n4 = {1, 2, 3, 4};
    assert((nextGreaterElement(n3, n4) == vector<int>{3, -1}));

    cout << "All tests passed!" << endl;
    return 0;
}
